onyo/lib/onyo.py

```diff
@@ -92,6 +92,94 @@ def clear_caches(self) -> None:
+                                keys: list[str] = [], destination: str = "") -> str:
+        """
+        If `message` is given, the function uses the first element of it to
+        generate a message subject, and the following ones for the message body.
+
+        If no `message` is given, generates a default commit message based on
+        the other parameters given, and the `files staged` in the repository.
+        The paths for the message are shortened if needed, so that the resulting
+        message subject does not exceed `MAX_LEN`.
+
+        Adds a list of `changed files` with there path relative to the root of
+        the repository to the body of all commit messages.
+        """
+        message_subject = ""
+        message_body = ""
+        message_appendix = ""
+        MAX_LEN = 80
+
+        # staged files and directories (without ".anchor") in alphabetical order
+        staged_changes = [Path(self.opdir, x).relative_to(self.root).parent
+                          if str(Path(self.opdir, x).relative_to(self.root).name) == ".anchor"
+                          else Path(self.opdir, x).relative_to(self.root)
+                          for x in sorted(self.files_staged)]
+
+        if message:
+            message_subject = message[0][0]
+            message_body = self._n_join([x[0] for x in message[1:]])
+        else:
+            # get variables for the begin of the commit message `msg_dummy`
+            dest = None
+            keys_str = ""
+            if keys:
+                keys_str = f" ({','.join(str(x.split('=')[0]) for x in sorted(keys))})"
+            if destination:
+                dest = Path(self.opdir, destination).relative_to(self.root)
+            if dest and str(dest.name) == ".anchor":
+                dest = dest.parent
+
+            # build the commit message subject
+            msg_dummy = f"{cmd} [{len(staged_changes)}]{keys_str}"
+            message_subject = self._generate_commit_message_subject(msg_dummy,
+                                                                    staged_changes,
+                                                                    dest, MAX_LEN)
+
+        message_appendix = self._n_join([str(x) for x in staged_changes])
+        return f"{message_subject}\n\n{message_body}\n\n{message_appendix}"
+
+    def _generate_commit_message_subject(self, msg_dummy: str,
+                                         staged_changes: list[Path],
+                                         destination: Optional[Path],
+                                         MAX_LEN: int) -> str:
+        """
+        Generates "commit message subject" with the `msg_dummy` and the paths
+        from `staged_changes` and `destination`, and shortens the paths if the
+        message length exceeds the `MAX_LEN`.
+        """
+
+        # long message: full paths (relative to repo-root)
+        paths_str = ','.join(["'{}'".format(x) for x in staged_changes])
```

Comment:
Nit: This might be a bit more readable (and consistent) if it uses f-strings.

Reply:
Maybe, but I really don't know how to do that for the given list, to come to the result of them being all quoted and comma seperated for different lenghts :(

Reply:
Ahh, I didn't mean the entire thing, rather:

```
paths_str = ','.join([f"'{x}'" for x in staged_changes])
```